presto-hive/src/main/java/com/facebook/presto/hive/SortingFileWriterFactory.java

```diff
@@ -0,0 +1,108 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SortingFileWriterFactory
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final ConnectorSession session;
+    private final JobConf conf;
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final DataSize sortBufferSize;
+    private final int maxOpenSortFiles;
+    private final PageSorter pageSorter;
+    private final OrcFileWriterFactory orcFileWriterFactory;
+    private final boolean sortedWriteToTempPathEnabled;
+    private final int sortedWriteTempFileSubdirectoryCount;
+
+    public SortingFileWriterFactory(
+            HdfsEnvironment hdfsEnvironment,
+            ConnectorSession session,
+            JobConf conf,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders,
+            DataSize sortBufferSize,
+            int maxOpenSortFiles,
+            PageSorter pageSorter,
+            OrcFileWriterFactory orcFileWriterFactory,
+            boolean sortedWriteToTempPathEnabled,
+            int sortedWriteTempFileSubdirectoryCount)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.session = requireNonNull(session, "session is null");
+        this.conf = requireNonNull(conf, "conf is null");
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(requireNonNull(sortFields, "sortFields is null"));
+        this.sortOrders = ImmutableList.copyOf(requireNonNull(sortOrders, "sortOrders is null"));
+        this.sortBufferSize = requireNonNull(sortBufferSize, "sortBufferSize is null");
+        this.maxOpenSortFiles = maxOpenSortFiles;
+        this.pageSorter = requireNonNull(pageSorter, "pageSorter is null");
+        this.orcFileWriterFactory = requireNonNull(orcFileWriterFactory, "orcFileWriterFactory is null");
+        this.sortedWriteToTempPathEnabled = sortedWriteToTempPathEnabled;
+        this.sortedWriteTempFileSubdirectoryCount = sortedWriteTempFileSubdirectoryCount;
+    }
+
+    public SortingFileWriter createSortingFileWriter(Path path, HiveFileWriter outputWriter, int fileNumber, Optional<Path> tempPath)
+    {
+        checkState(tempPath.isPresent() == sortedWriteToTempPathEnabled, "tempPath existence is not consistent with sortedWriteToTempPathEnabled config");
+
+        FileSystem fileSystem;
+        try {
+            fileSystem = hdfsEnvironment.getFileSystem(session.getUser(), path, conf);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_WRITER_OPEN_ERROR, e);
+        }
+
+        Path prefix = sortedWriteToTempPathEnabled
+                ? new Path(tempPath.get(), format(".tmp-sort-%s/.tmp-sort-%s", fileNumber % sortedWriteTempFileSubdirectoryCount, path.getName()))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;%&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 93. Consider formatting this data with &#39;%d&#39; instead.